File: viewer-macos/Sources/AppControlViewer/Net/SignalingClient.swift
import Foundation

struct IceServerConfig: Codable {
    let urls: [String]
    let username: String?
    let credential: String?
}

/// WebSocket-Client zum Rendezvous-Server.
///
/// Der Server sieht ausschließlich die Raum-ID und undurchsichtige
/// Base64-Blöcke. Diese Klasse kennt daher keinerlei Anwendungslogik — sie
/// transportiert Bytes. Die Verschlüsselung passiert eine Schicht darüber im
/// `SessionCoordinator`. Siehe docs/04-protocol.md §4.1.
@MainActor
final class SignalingClient: NSObject {

    /// Ein verschlüsselter Umschlag von der Gegenseite (bereits Base64-dekodiert).
    var onEnvelope: ((Data) -> Void)?
    var onIceServers: (([IceServerConfig]) -> Void)?
    var onPeerJoined: (() -> Void)?
    var onPeerLeft: (() -> Void)?
    var onServerError: ((String, String) -> Void)?
    var onDisconnected: ((Error?) -> Void)?

    private var task: URLSessionWebSocketTask?
    private var session: URLSession?
    private(set) var isConnected = false

    func connect(to url: URL, roomId: String) {
        // delegateQueue: .main, damit alle Callbacks auf dem Main-Actor landen
        // und die UI ohne zusätzliches Dispatching aktualisiert werden kann.
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 15
        session = URLSession(configuration: configuration, delegate: self, delegateQueue: .main)

        task = session?.webSocketTask(with: url)
        task?.resume()
        isConnected = true

        send(json: ["t": "join", "roomId": roomId, "role": "viewer", "v": "ac/1"])
        receiveNext()
        startKeepAlive()
    }

    /// Sendet einen verschlüsselten Umschlag an die Gegenseite.
    func sendEnvelope(_ envelope: Data) {
        send(json: ["t": "relay", "payload": envelope.base64EncodedString()])
    }

    func disconnect() {
        send(json: ["t": "leave"])
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
        session?.invalidateAndCancel()
        session = nil
        isConnected = false
    }

    // ── Intern ───────────────────────────────────────────────────────────────

    private func send(json object: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: object),
              let text = String(data: data, encoding: .utf8) else { return }

        task?.send(.string(text)) { [weak self] error in
            guard let error else { return }
            Task { @MainActor in
                self?.isConnected = false
                self?.onDisconnected?(error)
            }
        }
    }

    private func receiveNext() {
        task?.receive { [weak self] result in
            Task { @MainActor in
                guard let self else { return }

                switch result {
                case .success(let message):
                    switch message {
                    case .string(let text):
                        self.handle(text)
                    case .data(let data):
                        // Der Server sendet ausschließlich Text. Binärframes sind
                        // ein Protokollfehler — verwerfen statt raten.
                        NSLog("AppControl: unerwarteter Binärframe (%d Byte) verworfen", data.count)
                    @unknown default:
                        break
                    }
                    self.receiveNext()   // nächste Nachricht anfordern

                case .failure(let error):
                    self.isConnected = false
                    self.onDisconnected?(error)
                }
            }
        }
    }

    private func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = object["t"] as? String else { return }

        switch type {
        case "relay":
            if let payload = object["payload"] as? String,
               let raw = Data(base64Encoded: payload) {
                onEnvelope?(raw)
            }

        case "joined":
            if let peerPresent = object["peerPresent"] as? Bool, peerPresent {
                onPeerJoined?()
            }

        case "peer-joined":
            onPeerJoined?()

        case "peer-left":
            onPeerLeft?()

        case "ice-servers":
            if let servers = object["servers"],
               let serverData = try? JSONSerialization.data(withJSONObject: servers),
               let parsed = try? JSONDecoder().decode([IceServerConfig].self, from: serverData) {
                onIceServers?(parsed)
            }

        case "error":
            onServerError?(
                object["code"] as? String ?? "unknown",
                object["message"] as? String ?? "")

        default:
            break   // Vorwärtskompatibilität: unbekannte Typen ignorieren
        }
    }

    /// WebSocket-Ping alle 20 s. Ohne ihn schließen NAT-Router und Load-Balancer
    /// eine still daliegende Verbindung nach typischerweise 30–60 s.
    private func startKeepAlive() {
        Task { @MainActor [weak self] in
            while let self, self.isConnected {
                try? await Task.sleep(for: .seconds(20))
                guard self.isConnected else { break }
                self.task?.sendPing { _ in }
            }
        }
    }
}

extension SignalingClient: URLSessionWebSocketDelegate {
    nonisolated func urlSession(_ session: URLSession,
                                webSocketTask: URLSessionWebSocketTask,
                                didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
                                reason: Data?) {
        Task { @MainActor [weak self] in
            self?.isConnected = false
            self?.onDisconnected?(nil)
        }
    }
}

File: host-windows/src/AppControl.Host/Net/SignalingClient.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;

namespace AppControl.Host.Net;

public sealed record IceServerConfig
{
    [JsonPropertyName("urls")]       public string[] Urls { get; init; } = [];
    [JsonPropertyName("username")]   public string? Username { get; init; }
    [JsonPropertyName("credential")] public string? Credential { get; init; }
}

/// <summary>
/// WebSocket-Client zum Rendezvous-Server.
///
/// Der Server sieht ausschliesslich die Raum-ID und undurchsichtige Base64-Bloecke.
/// Diese Klasse kennt daher keinerlei Anwendungslogik - sie transportiert Bytes.
/// Die Verschluesselung passiert eine Schicht darueber in SessionController.
/// Siehe docs/04-protocol.md §4.1.
/// </summary>
public sealed class SignalingClient : IAsyncDisposable
{
    private readonly ILogger<SignalingClient> _log;
    private ClientWebSocket? _socket;
    private CancellationTokenSource? _cts;
    private Task? _receiveLoop;

    /// <summary>Ein verschluesselter Umschlag von der Gegenseite (bereits Base64-dekodiert).</summary>
    public event Action<byte[]>? EnvelopeReceived;

    public event Action<IceServerConfig[]>? IceServersReceived;
    public event Action? PeerJoined;
    public event Action? PeerLeft;
    public event Action<string, string>? ServerError;
    public event Action<Exception?>? Disconnected;

    public SignalingClient(ILogger<SignalingClient> log) => _log = log;

    public bool IsConnected => _socket?.State == WebSocketState.Open;

    public async Task ConnectAsync(Uri serverUri, string roomId, CancellationToken ct = default)
    {
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        _socket = new ClientWebSocket();
        _socket.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);

        await _socket.ConnectAsync(serverUri, _cts.Token);
        _log.LogInformation("Mit Signaling-Server verbunden: {Uri}", serverUri);

        await SendJsonAsync(new { t = "join", roomId, role = "host", v = "ac/1" }, _cts.Token);

        _receiveLoop = Task.Run(() => ReceiveLoopAsync(_cts.Token), _cts.Token);
    }

    /// <summary>Sendet einen verschluesselten Umschlag an die Gegenseite.</summary>
    public Task SendEnvelopeAsync(ReadOnlyMemory<byte> envelope, CancellationToken ct = default)
        => SendJsonAsync(new { t = "relay", payload = Convert.ToBase64String(envelope.Span) }, ct);

    private async Task SendJsonAsync(object message, CancellationToken ct)
    {
        if (_socket is not { State: WebSocketState.Open })
            throw new InvalidOperationException("Signaling-Verbindung ist nicht offen");

        var json = JsonSerializer.SerializeToUtf8Bytes(message);
        await _socket.SendAsync(json, WebSocketMessageType.Text, endOfMessage: true, ct);
    }

    private async Task ReceiveLoopAsync(CancellationToken ct)
    {
        // 256 KiB: Ein SDP mit vielen ICE-Kandidaten kann mehrere zehn KiB gross
        // werden; der Server begrenzt bei 128 KiB Base64.
        var buffer = new byte[256 * 1024];
        Exception? failure = null;

        try
        {
            while (!ct.IsCancellationRequested && _socket?.State == WebSocketState.Open)
            {
                var received = 0;
                WebSocketReceiveResult result;
                do
                {
                    result = await _socket.ReceiveAsync(
                        new ArraySegment<byte>(buffer, received, buffer.Length - received), ct);

                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        _log.LogInformation("Signaling-Server hat die Verbindung geschlossen: {Status} {Desc}",
                            result.CloseStatus, result.CloseStatusDescription);
                        return;
                    }
                    received += result.Count;
                }
                while (!result.EndOfMessage && received < buffer.Length);

                HandleMessage(Encoding.UTF8.GetString(buffer, 0, received));
            }
        }
        catch (OperationCanceledException) { /* normaler Abbau */ }
        catch (Exception ex)
        {
            failure = ex;
            _log.LogError(ex, "Signaling-Verbindung abgebrochen");
        }
        finally
        {
            Disconnected?.Invoke(failure);
        }
    }

    private void HandleMessage(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (!root.TryGetProperty("t", out var typeProp)) return;

            switch (typeProp.GetString())
            {
                case "relay":
                    if (root.TryGetProperty("payload", out var payload))
                    {
                        var raw = Convert.FromBase64String(payload.GetString() ?? "");
                        EnvelopeReceived?.Invoke(raw);
                    }
                    break;

                case "joined":
                    _log.LogInformation("Raum betreten, Gegenpart anwesend: {Present}",
                        root.TryGetProperty("peerPresent", out var p) && p.GetBoolean());
                    if (root.TryGetProperty("peerPresent", out var pp) && pp.GetBoolean())
                        PeerJoined?.Invoke();
                    break;

                case "peer-joined": PeerJoined?.Invoke(); break;
                case "peer-left":   PeerLeft?.Invoke();   break;

                case "ice-servers":
                    if (root.TryGetProperty("servers", out var servers))
                    {
                        var parsed = servers.Deserialize<IceServerConfig[]>() ?? [];
                        _log.LogInformation("{Count} ICE-Server erhalten", parsed.Length);
                        IceServersReceived?.Invoke(parsed);
                    }
                    break;

                case "error":
                    var code = root.TryGetProperty("code", out var c) ? c.GetString() ?? "?" : "?";
                    var msg = root.TryGetProperty("message", out var m) ? m.GetString() ?? "" : "";
                    _log.LogWarning("Signaling-Fehler {Code}: {Message}", code, msg);
                    ServerError?.Invoke(code, msg);
                    break;
            }
        }
        catch (Exception ex)
        {
            // Kaputtes JSON vom Server ist kein Grund abzustuerzen - aber ein
            // Grund, es sichtbar zu machen.
            _log.LogWarning(ex, "Unlesbare Nachricht vom Signaling-Server verworfen");
        }
    }

    public async ValueTask DisposeAsync()
    {
        try
        {
            if (_socket is { State: WebSocketState.Open })
            {
                await SendJsonAsync(new { t = "leave" }, CancellationToken.None);
                await _socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "bye", CancellationToken.None);
            }
        }
        catch { /* beim Abbau ist ein Fehler folgenlos */ }

        _cts?.Cancel();
        if (_receiveLoop is not null)
        {
            try { await _receiveLoop.WaitAsync(TimeSpan.FromSeconds(2)); } catch { }
        }
        _socket?.Dispose();
        _cts?.Dispose();
    }
}
